### ISSUE TITLE
What is a digest or reference of manifest?

### ISSUE THREAD

**Issue by least-olegs** [docker/distribution#1831]
The documentation is completely opaque when it describes deleting images:

> 
For deletes, reference must be a digest or the delete will fail. If the image exists and has been successfully deleted, the following response will be issued:

This is extremely unhelpful given how many places in Docker look like digests of something (strings of random numbers).

Whatever "digest" is, it is not included in response when retrieving a list of images. Also, it is possible to retrieve a manifest w/o specifying any "digests", so why is it needed here?

On top of all of this: why is the API user needs to know about how you manage the storage - manifests, blobs etc. - it's your internal business the user wants no part of. Users know images via names and tags - why suddenly do the user needs so much information when it comes to deleting?

**Comment by RichardScothern** [comment]
> This is extremely unhelpful given how many places in Docker look like digests of something (strings of random numbers).

Distribution digests are described in the API specification (https://docs.docker.com/registry/spec/api/#/content-digests)

> Whatever "digest" is, it is not included in response when retrieving a list of images. Also, it is possible to retrieve a manifest w/o specifying any "digests", so why is it needed here?

CAS is the cornerstone of distribution storage.  Tags are mutable descriptors.  The digest of an object is the only sure way of identifying that object.  The digest of an object is available via stat-ing an object.

> On top of all of this: why is the API user needs to know about how you manage the storage - manifests, blobs etc. - it's your internal business the user wants no part of. Users know images via names and tags - why suddenly do the user needs so much information when it comes to deleting?

If the user could not distinguish between a blob and a manifest, how would they differentiate between them for the purpose (amongst others) of deleting?  Digests as identifiers have a role beyond deleting.  Manifests form a merkle tree of content that is verifiable and unforgeable.  This information is used to build trust systems.

**Comment by least-olegs** [comment]
Imagine that... I did, in fact, read the section on "digests", and yes, this is why I say that you put lots of sequences of random numbers everywhere. When you use the word "digest" to mean something only you know the meaning of, no one else is going to know what particular sequence of random numbers you are referring to.

What are you talking about? Why does the user needs to distinguish between blobs and manifests if the user isn't even supposed to *know* any of those things exist? This is the same as saying "you cannot delete a file using file name only, instead, you must provide file name and inode number because file names don't identify blocks uniquely" - who cares? This is not the level of API you should be thinking at.

Can you please be at least a little bit helpful? I am not using Docker because I like it - I have to, because the company decided to. You don't need to take my word for it, when I tell you the API is unusable, you don't need to change anything, if you disagree. Just, please, tell me how to use it - there's no way an average person can figure this out from your description.

**Comment by least-olegs** [comment]
Alright, substituting everything that looked like a digest into the API calls finally uncovered the one that I had to use. Being a prick as I am, I will not write where and how this was discovered, instead, I'll let you deal with more puzzled inquires to come--in my experience, if workaround is offered, then the developer will not fix the problem as promptly.

**Comment by RichardScothern** [comment]
A digest is specified in the document I linked.  It is not a sequence of random numbers - it has structure and is verifiable. 

The API for the registry exposes blob and manifests.   Any client interacting with the registry needs to know these concepts.  The registry is a system level component, who's main client is the docker engine.  For most users, that is the point of interaction.  If you are interacting directly with the registry API you will need some knowledge of distribution concepts.

I was trying to be helpful.  Your issue was a lack of understanding around the registry API and I tried to help you with that.  If you had a more specific question ("How do I get a digest to delete an image"), I would have answered that.

Previous puzzled inquiries about documentation have often come with pull requests to improve it.  This is how open source projects work.  It is your choice to do or not do this.

**Comment by BrendonW** [comment]
Ok, so I give up. How do I get the digest of a manifest so that I can delete it?

As far as I can tell, the docker executable interaction with the registry does not give me a way to delete a tag that is pushed accidently. I really don't want to police people who make the same typo that I did and end up using the wrong thing.  I can use curl to get the manifest in question, but that does not include anything labeled "digest" and the curl command do delete the manifest wants the digest for the manifest.

The other gentleman is correct, many very simple commands are very hard to decode without lots of searching and trying things out. For something like docker that is poised to take over the world, that seems like a very sad state of affairs -- and the reason everybody (Google/Amazon/etc.) is creating their own "almost compatible" tools...

**Comment by RichardScothern** [comment]
hello @BrendonW .  When you curl the manifest, the digest is included in the header.  This is documented in the API reference in the manifest section: https://docs.docker.com/registry/spec/api/#/pulling-an-image

Now, obviously there is some deficiency in the documentation given that you and @least-olegs are having such problems finding information.  As a user you will be more aware of this than the maintainers who created it.   

In what way can this be improved?  Do we need a tutorial?  Are you willing to contribute back to the project to help improve matters?

**Comment by BrendonW** [comment]
@RichardScothern Thanks! 

That information would have saved me about 3 hours of messing around before I logged into the registry and just deleted the directory for the tag I needed to get rid of! (I think it worked because when I garbage collected, it found 1 item to collect, which was what I expected.) Since I wasn't interested in pulling an image, I didn't think to look in that part of the documentation. I missed the detail about the digest being in the header because all the rest of the information seems to be in the JSON itself.

I'd imagine some simple & complete examples might help. The bits that keep tripping me up are when some critical detail is assumed by the author, but I don't know either what 'it' is or where to find it. I'm keeping notes and was thinking about doing a blog post about my findings, so perhaps I should rather just help with the documentation? I can see the challenge of being too knowledgeable to create great documentation for beginners. What is the best way to start helping?

**Comment by RichardScothern** [comment]
A documentation PR would be the most beneficial.  It's the missing critical details I'm most interested in.  I think all of the information is in the spec (docs/spec/api.md), but people don't want to wade through it.  Perhaps distilling some of the information there with some clarifications into a tutorial / walkthrough would help?

**Comment by BrendonW** [comment]
Ok, I'll work on that as time allows. I've still not managed to delete
using the v2 http api... I did enable delete, which is the 2nd gotcha.

I now can get the digest from the header:

> curl -I "http://localhost:5000/v2/ruby-base/manifests/v1"

...

Docker-Content-Digest:
> sha256:3fdc868cf1f86b299bd3284e4eb302d3849c420c4852f67ff8541e090377fa67


but when I try to issue the manifest delete:

> curl -i -X DELETE "
> http://localhost:5000/v2/ruby-base/manifests/3fdc868cf1f86b299bd3284e4eb302d3849c420c4852f67ff8541e090377fa67
> "
>
I get the response 400 Bad Request

> {"errors":[{"code":"DIGEST_INVALID","message":"provided digest did not
> match uploaded content"}]}


I'll have to figure that out later, because it sure seems to match!
- Brendon

On Fri, Jul 22, 2016 at 10:07 AM, Richard Scothern <notifications@github.com

**Comment by RichardScothern** [comment]
TLDR: there is an accept header required to prevent the registry rewriting the manifest which would give the incorrect digest.

https://docs.docker.com/registry/spec/api/#/deleting-an-image

**Comment by BrendonW** [comment]
Thanks, I can now delete. Much more like Read a bunch of stuff that mislead
me into the wrong conclusion and missed a few detail clues that could have
got me back on track.

Delete really does not do what I expected. I wanted to delete an incorrect
tag and it zapped every version of the image.

A couple of details that tripped me up the first time:

   - The digest must include the "sha256:" prefix
   - The delete deletes ALL entries of the same image. (The digest isn't of
   the manifest, it seems to be of the image.) So if you push a version with
   the wrong tag, which was my problem, then you end up having to remember to
   download the image before doing the delete so that you can re-push all the
   correct tags!
   - Possibly not surprising, but the accept header changes the manifest
   format completely.

The documentation is a bit confusing:

> IMPORTANT: If a *digest* is used to fetch content, the client should use
> the same digest used to fetch the content to verify it. The header
> Docker-Content-Digest should not be trusted over the “local” digest.


I don't even know what that is trying to say. "I should verify the digest,
but should not trust the digest?"

And from the compatibility doc, it says the following:

> The docker engine will construct a *Schema 2* manifest which the registry
> will persist to disk.
> When the manifest is pulled by digest or tag with Docker Engine 1.10, a *Schema
> 2* manifest will be returned. The Docker Engine 1.10 understands the new
> manifest format.
> When the manifest is pulled by *tag* with Docker Engine 1.9 and older,
> the manifest is converted on-the-fly to *Schema 1*and sent in the
> response. The Docker Engine 1.9 is compatible with this older format.
> *When the manifest is pulled by digest with Docker Engine 1.9 and older,
> the same rewriting process will not happen in the registry. If this were to
> happen the digest would no longer match the hash of the manifest and would
> violate the constraints of CAS.*

For this reason if a manifest is pulled by *digest* from a registry 2.3
> with Docker Engine 1.9 and older, and the manifest was pushed with Docker
> Engine 1.10, a security check will cause the Engine to receive a manifest
> it cannot use and the pull will fail.


Since I'm using a 2.5.0 registry and a 1.12.0 Engine, this part of the docs
led me to believe that I would just get the right header and not need the
accept header to get the right value.

TLDR; it's easier just to delete the tag-manifest directly from the
registry filesystem.

- Brendon

On Mon, Jul 25, 2016 at 11:19 AM, Richard Scothern <notifications@github.com

**Comment by lawliet89** [comment]
I am equally having issues deleting a manifest. Let's say I am trying to delete the (only) tag of the image `oracle-xe-11g`. 

Firstly:

```
$ bin/registry garbage-collect --dry-run /etc/docker/registry/config.yml
...
oracle-xe-11g
oracle-xe-11g: marking manifest sha256:4df87a33896841bcbe24ede88434d00d7eb76122f2d79559049208968f05de5c 
oracle-xe-11g: marking blob sha256:0653bff3c5cf23727e0ebceae7a28f7534ab64ed13966e080e4c9b035176c401
oracle-xe-11g: marking blob sha256:8cc9e686e4090db851ed55108fbf1f99f684a7826e296139d582d650b42fe14a
oracle-xe-11g: marking blob sha256:21d38d6f9835a5b92682a1e54951948df92cb843e10a40ff12a9f0b487177c6a
oracle-xe-11g: marking blob sha256:12767b5ff7d9d9b4f43100ec2ab78866f3173803f73faad008466d26d811465c
oracle-xe-11g: marking blob sha256:552483024f6514d0f60acc7d995a3e6aece729f21ffce9ee3cc18166f5673b79
oracle-xe-11g: marking configuration sha256:51c3a703fa3ad214fa4ec201641a6e4446da17986740f8cbed9308cff728c254
oracle-xe-11g: marking manifest sha256:f18fbad54f146e89e66de37ec05ee03e53bafac647bc9ed8a1acbc8df08aa177 
oracle-xe-11g: marking blob sha256:0653bff3c5cf23727e0ebceae7a28f7534ab64ed13966e080e4c9b035176c401
oracle-xe-11g: marking blob sha256:8cc9e686e4090db851ed55108fbf1f99f684a7826e296139d582d650b42fe14a
oracle-xe-11g: marking blob sha256:21d38d6f9835a5b92682a1e54951948df92cb843e10a40ff12a9f0b487177c6a
oracle-xe-11g: marking blob sha256:12767b5ff7d9d9b4f43100ec2ab78866f3173803f73faad008466d26d811465c
oracle-xe-11g: marking blob sha256:552483024f6514d0f60acc7d995a3e6aece729f21ffce9ee3cc18166f5673b79
oracle-xe-11g: marking configuration sha256:1601ac4324b6753ade84f47423295c3d85c74ed57a57e9c6eeb668fb21727697
...
```
But when I `GET https://docker-registry.example.com/v2/oracle-xe-11g/manifests/latest`, I get header `Docker-Content-Digest: sha256:27baa00c4ff8067257c6198530a6cf11438f79b363caccebfeebfbe13208f8e1`

What is going on? The header is nowhere to be found from the console output above. Obviously if I do `DELETE https://docker-registry.example.com/v2/oracle-xe-11g/manifests/sha256:27baa00c4ff8067257c6198530a6cf11438f79b363caccebfeebfbe13208f8e1`, I get 

```json
{
  "errors": [
    {
      "code": "MANIFEST_UNKNOWN",
      "message": "manifest unknown"
    }
  ]
}
```

But it works if I use the manifests from the console output above.

**Comment by matutter** [comment]
@RichardScothern, thanks for being so responsive all over your project's issues, I've spent about 2 days getting my head around docker engine/registry API and your posts on here usually clear up everything.

Deleting an image has been a particularly difficult part of the registry API. 

> In what way can this be improved? Do we need a tutorial? Are you willing to contribute back to the project to help improve matters? 

I don't think a tutorial would be necessary, usually they just confuse people who find them when the next version comes out. But I do think the documentation, specifically for the registry/spec does need some fine tuning and I have just 3 points on that.

1. For an API to use different HTTP features such as header/query-string/or message body information for each endpoint should be visually distinct either by organization or color.
Example, [deleting-an-image](https://docs.docker.com/registry/spec/api/#/deleting-an-image) first shows an example of the API call, a ```DELETE``` to ```/v2/<name>/manifests/<reference>``` then shows me that a "success" response is a status-code ```202``` Then a section that is designated ```Note``` tells us about the HTTP headers. This should be in a section clearly marked headers. Also not all sections follow this sequence of information.

2. Each endpoint should have every HTTP feature documented, even if an endpoint doesn't use said feature, be it QS, headers, or body. 

3. Many times when there is an endpoint example such as ```DELETE /v2/<name>/manifests/<reference>``` the reality is that "reference" may be a tag or digest. I think your Docker-Engine API docs do a better job at this, for example ```GET /containers/(id or name)/json``` 

For what it's worth I think the [docker-remote-api](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/) is written wonderfully and that format would be a great benefit to this project.

**Comment by BrendonW** [comment]
I think the biggest confusion I had, was that when you do requests
incorrectly, they look exceedingly similar to the correct calls, especially
when the difference is just in the headers. Another is that you can
retrieve reference hash values that "make sense" but are not the ones that
the delete command is expecting.

That sets you up with code that looks identical to the examples, but just
doesn't work. -- To be clear, the code isn't identical, the headers need to
be changed, but that info is buried in the documentation. If should be
front&center!

- Brendon

**Comment by RichardScothern** [comment]
hello @matutter 

Thanks for the constructive criticism, I find your comment genuinely useful.

I think taking on the format of the docker remote API is a good idea.  Having a uniform REST API documentation format makes perfect sense.  I will talk to our documentation team to see how this is generated and see if we can do the same for distribution.

**Comment by ebd2** [comment]
@RichardScothern: I also had trouble sorting through this documentation to implement tag deletion correctly. I read the docs, the thread on this issue, and this [stackoverflow](http://stackoverflow.com/questions/37033055/how-can-i-use-the-docker-registry-api-v2-to-delete-an-image-from-a-private-regis/39983191) post, and I still haven't gotten deletion to work against dockerhub.

I'm getting the error UNSUPPORTED, which isn't completely informative: If I pass a completely bogus digest when I make the DELETE request, I get the same error. At this point, I'm not sure if I've read the docs incorrectly, or if deletion via the registry v2 API is genuinely unsupported on dockerhub via the registry API.

If either you or @BrendonW, who got deletion working, is willing to take a quick look at my code, it's up [here](https://github.com/ebd2/dreg-cleaner). The code is somewhat specific to our particular workflow, but hopefully it's clear enough to follow. That said, it is hackathon code, and I'm not primarily a Python programmer.

**Comment by RichardScothern** [comment]
@ebd2 delete is not enabled on the hub registry.  Deletion is handled by hub-specific endpoints.

**Comment by BrendonW** [comment]
At a glance, the code seems OK Eric. There are three things that I've seen
people, including myself do wrong:

   1. The registry doesn't have storage: delete: enabled: true set in the
   config.yml.
   2. You don't get the correct <digest> for the thing you want to delete.
   The correct one is returned in the header to the
   /v2/<name>/manifest/<tag> request.
   3. Not using the DELETE verb in the delete request.

It would be nice if the error message gave you a more helpful clue on
failure. "Unknown" vs "Not Allowed" would provide a great clue :)

Eric, the easy way to test this is to pull down a docker registry image and
test locally.

- Brendon

**Comment by ebd2** [comment]
Thanks @RichardScothern and @BrendonW for taking a look. Based on Richard's response it sounds like the issue is number 1 or something analogous to it on the public docker hub.

I'm in agreement that maybe a 404 might be a useful error code to return if you call the DELETE endpoint on a digest that doesn't actually exist and using 405 for the case of the registry not supporting delete.

Sorry for the thread-jack :-(

**Comment by lazize** [comment]
I just got the correct clue about what is the "correct" digest reading this issue, I couldn't find it in the documentation, and I read it many times.

Let me make it clear for anyone will read it in the future:

This is **WRONG**:
```
curl -XGET -H "Accept: application/vnd.docker.distribution.manifest.v2+json" https://127.0.0.1:5000/v2/ubuntu-leo/manifests/latest
```

This is THE correct way to get the digest:
```
curl --head -XGET -H "Accept: application/vnd.docker.distribution.manifest.v2+json" https://127.0.0.1:5000/v2/ubuntu-leo/manifests/latest
```

You MUST use the option `--head` or `-I` in your request to the the correct digest value.

That is the output with the correct digest, called `Docker-Content-Digest`:
```
HTTP/1.1 200 OK
Content-Length: 1357
Content-Type: application/vnd.docker.distribution.manifest.v2+json
Docker-Content-Digest: sha256:634a341aa83f32b48949ef428db8fefcd897dbacfdac26f044b60c14d1b5e972
Docker-Distribution-Api-Version: registry/2.0
Etag: "sha256:634a341aa83f32b48949ef428db8fefcd897dbacfdac26f044b60c14d1b5e972"
X-Content-Type-Options: nosniff
Date: Wed, 16 Aug 2017 14:55:07 GMT
```

So, to delete the command will be like below, as explained above in this issue, `sha256:` is part of the digest value.
```
curl -XDELETE https://127.0.0.1:5000/v2/ubuntu-leo/manifests/sha256:634a341aa83f32b48949ef428db8fefcd897dbacfdac26f044b60c14d1b5e972
```

**Comment by lazize** [comment]
As @BrendonW explained it will not delete just one tag, it will delete everything from that specific repository.

Is there a way to delete just one tag?
If I want to keep just the last 10 tag's, how can I do that? Do I need to pull all 10 tags, delete with the instruction mentioned above and push back all 10 tags?

If this issue is not the correct one for this question, I can open a new one and reference this, just let me know!

**Comment by younkim** [comment]
Not sure if there has been an update regarding this in Docker registry, but when you push or pull to the registry, I see the digest hash. Is this new?

@lazize Any idea where I'm supposed to place the configuration yml file for the registry? the delete option needs to be enabled explicitly.

**Comment by adamdecaf** [comment]
I ran into this (and https://github.com/docker/distribution/issues/1566#issuecomment-224999363) and had a big amount of trouble getting it to work.

Docs would be nice.
